src/render/image_atlas.js

```diff
@@ -83,6 +92,10 @@ export default class ImageAtlas {
             };
             bins.push(bin);
             patternPositions[id] = new ImagePosition(bin, src);
+
+            if (src.hasRenderCallback) {
+                haveRenderCallbacks.push(id);
+            }
```

Comment:
These two blocks (icons and patterns loop) start to become too repetitive. Let's cut this out into a helper function to DRY it up.